tests/MongoDB.Driver.Tests/JsonDrivenTests/JsonDrivenWithTransactionTest.cs

```diff
@@ -0,0 +1,150 @@
+﻿/* Copyright 2019-present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using FluentAssertions;
+using MongoDB.Bson;
+using MongoDB.Bson.TestHelpers.JsonDrivenTests;
+
+namespace MongoDB.Driver.Tests.JsonDrivenTests
+{
+    public sealed class JsonDrivenWithTransactionTest : JsonDrivenSessionTest
+    {
+        // private fields
+        private TransactionOptions _options = null;
+        private Func<IClientSessionHandle, CancellationToken, BsonDocument> _callback = null;
+        private Func<IClientSessionHandle, CancellationToken, Task<BsonDocument>> _callbackAsync = null;
+        private readonly JsonDrivenTestFactory _jsonDrivenTestFactory;
+        private BsonDocument _result;
+        private const string ErrorFlag = "errorLabelsOmit";
+
+        // public constructors
+        public JsonDrivenWithTransactionTest(JsonDrivenTestFactory jsonDrivenTestFactory, Dictionary<string, object> objectMap) : base(objectMap)
+        {
+            _jsonDrivenTestFactory = jsonDrivenTestFactory;
+        }
+
+        // public methods
+        public override void Arrange(BsonDocument document)
+        {
+            JsonDrivenHelper.EnsureAllFieldsAreValid(document, "name", "object", "arguments", "result");
+
+            base.Arrange(document);
+        }
+
+        // protected methods
+        protected override void AssertResult()
+        {
+            _result.Should().ShouldBeEquivalentTo(_expectedResult.AsBsonDocument);
+        }
+
+        protected override void CallMethod(CancellationToken cancellationToken)
+        {
+            _result = _session.WithTransaction(_callback, _options, cancellationToken);
+        }
+
+        protected override async Task CallMethodAsync(CancellationToken cancellationToken)
+        {
+            _result = await _session.WithTransactionAsync(_callbackAsync, _options, cancellationToken).ConfigureAwait(false);
+        }
+
+        protected override void SetArgument(string name, BsonValue value)
+        {
+            switch (name)
+            {
+                case "options":
+                    _options = ParseOptions(value.AsBsonDocument);
+                    return;
+                case "callback":
+                    {
+                        var valueBson = value.AsBsonDocument;
```

Comment:
The normal suffix for a variable that is a BsonDocument is "Document":

    var valueDocument = value.AsBsonDocument;

Reply:
Done